dao/src/main/java/se/fortnox/reactivewizard/db/transactions/Transaction.java

```diff
@@ -94,16 +99,25 @@ private void closeConnection(Connection connection) {
         }
     }
 
-    private void allFailed(Connection connection, Throwable throwable) {
-        for (TransactionStatement statement : statementsToExecute) {
+    private void allFailed(Throwable throwable) {
+        for (TransactionStatement transactionStatement : statementsToExecute) {
+            final Statement statement = transactionStatement.getAtomicStatement().get();
             try {
-                statement.getStatement().onError(throwable);
+                setTransactionToExecutable(transactionStatement, statement);
+                statement.onError(throwable);
             } catch (Exception onErrorException) {
                 log.error("onError threw exception", onErrorException);
             }
         }
     }
 
+    private void setTransactionToExecutable(TransactionStatement transactionStatement, Statement statement) {
+        final AtomicReference<Statement> atomicStatement = transactionStatement.getAtomicStatement();
+        final Transaction                transaction     = transactionStatement.getTransaction();
+        atomicStatement.compareAndSet(statement,null);
+        transaction.getWaitingForExecution().compareAndSet(false,true);
```

Comment:
Why compare and set if you are not using the result? Why doing this for each statement?